providers/chefinfra/chefinfra.go

```diff
@@ -16,12 +16,101 @@ limitations under the License.
 
 package chefinfra
 
-import "fmt"
+import (
+	"crypto/sha256"
+	"encoding/json"
+	"errors"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"os/exec"
+	"strings"
+)
 
 // func executePolicyArchive() {}
 // func passInAttrAsJson() {}
 
-// Execute execs
-func Execute() {
-	fmt.Println("Infra!")
+type chefInfraDefinition struct {
+	args           []string
+	policyURL      string `json:"policy_url"`
+	policySum      string `json:"policy_sum"`
+	jsonAttributes string `json:"attributes"`
+}
+
+// New creates a new Chef Infra Definition that can be used to execute Chef Infra Client.
+func New(m map[string]interface{}) (*chefInfraDefinition, error) {
+	chef := chefInfraDefinition(m)
+	path, err := exec.LookPath("chef-client")
+	if err != nil {
+		return nil, err
+	}
+
+	append(chef.args, path)
+	return chef, nil
+}
+
+// Prepare prepares the chef-infra-client commands and downloads and processes the archives.
+func (chef *chefInfraDefinition) Prepare() error {
+	if chef.policyURL != nil {
+		append(chef.args, "--local-mode")
+
+		s := strings.Split(chef.policyURL, "/")
+		filepath := ".foodtruck/chefinfra/" + s[len(s)-1]
+		append(chef.args, "--recipe-url", filepath)
+
+		err := downloadFile(chef.policyURL, filepath)
+		if err != nil {
+			return err
+		}
+	}
+	if chef.jsonAttributes != nil {
+		append(chef.args, "--json-attributes", ".foodtruck/chefinfra/dna.json")
```

Comment:
Do we use `append` instead of a pointer/deference here because chef.args is part of a struct?

Reply:
Or is it because we are passing it in as a deference, we then don't need to deference it again?

Reply:
append appends arguments to the end of a slice (similar to array.append() in other languages).  In this case, it's taking the string slice named "chef.args" and appending those strings to it.

I did it this way, because the exec.Command() function wants a slice of args to form the command.